### ISSUE TITLE
Allow async access control functions

### ISSUE THREAD

**Issue by jesstelford** [keystonejs/keystone#2358]
# Feature request

For the `create` access control on a list, you [can specify a function which returns `true`/`false`](https://www.keystonejs.com/api/access-control#granular-functions-returning-graphqlwhere).

Ideally we'd be able to also return a `GraphQLWhere` clause, but because the data doesn't yet exist for a `create`, that filtering isn't possible:

```javascript
// Only allow group owners to create events
// DOES NOT WORK; Known limitation
keystone.createList('Event', {
  access: {
    create: ({ authentication: { item, listKey } }) => ({
      group: { owners_some: { id: item.id } }
    }),
  }
});
```

Alternatively, we could do the same functionality by looking up the data ourselves while meeting the requirement of returning `true`/`false`

```javascript
keystone.createList('Event', {
  access: {
    create: async ({ authentication: { item }, originalInput }) => {
      // Must have the group ID set, if not; don't allow creating
      if (!originalInput.group || !originalInput.group.id) { return false }
      // Fetch all the owners for the group they've set
      const owners = await getOwnersOfGroup(originalInput.group.id);
      // If the current user is an owner, let them through. Otherwise, don't.
      return !!owners.find(({ id }) => id === item.id);
    },
});
```

However, this does not work as the access control checks require a synchronous value to be returned, not a `Promise`:

https://github.com/keystonejs/keystone/blob/7b4ed3623f5774d7783c39962bfa1ce97938e310/packages/access-control/lib/access-control.js#L154-L170

## Describe the solution you'd like

Allow returning a `Promise` from access control functions which must resolve to either `true`/`false` or a `GraphQLWhere` clause.

## Describe alternatives you've considered

One alternative is to actually implement `GraphQLWhere` support for `create` - we'd have to build out the functionality to do all the filtering on in-memory data. This would be 100% possible once the Memory Database Adapter is complete: #1310

**Comment by zenflow** [comment]
If only the access control functions were just passed the data of the item being operated on, that would help us in 95% of the use-cases and you wouldn't need this feature.

**Comment by wcalebgray** [comment]
@jesstelford what would it take to get this implemented? I think it's critical to an app that I'm building and am willing to look into it, but I have no idea where to get started. Alternatively, where is it on your guys's roadmap currently?

**Comment by jesstelford** [comment]
Start within the linked code:

https://github.com/keystonejs/keystone/blob/7b4ed3623f5774d7783c39962bfa1ce97938e310/packages/access-control/lib/access-control.js#L154-L170

Specifically, line 160 there should be changed:

```diff
-     result = access({ authentication: authentication.item ? authentication : {} }); 
+     result = await access({ authentication: authentication.item ? authentication : {} }); 
```

That then means the function `validateCustomAccessControl` becomes `async`:

```diff
- validateCustomAccessControl({ access, authentication = {} }) { 
+ async validateCustomAccessControl({ access, authentication = {} }) { 
```

Which means anywhere that _calls_ `validateCustomAccessControl` needs to also be `async`/`await`, and so-on up the call stack. I'm not sure if this is an easy task or not, and unfortunately don't have time to investigate right now.

In terms of priority; this is probably a lower priority for us as we haven't had any other requests for this feature beyond my own your just now.

More than happy to assist with questions as you implement! Thank you <3

**Comment by wcalebgray** [comment]
Yeah, that's exactly where I was starting. Looks like it would require the following:

The `_prepare` method here is an `async` call already, so this is what we need to "bridge" to.
https://github.com/keystonejs/keystone/blob/da1359dfc1bff7e27505eff876efe3a0865bae2d/packages/keystone/lib/Keystone/index.js#L622

Move the `this.getGraphQlContext` call out of the `this._buildQueryHelper()` call. Make it an `await` call and assign it to a `const` that then is passed in to `this._buildQueryHelper()`.
https://github.com/keystonejs/keystone/blob/da1359dfc1bff7e27505eff876efe3a0865bae2d/packages/keystone/lib/Keystone/index.js#L630-L636

Make `getGraphQlContext` an `async` function
https://github.com/keystonejs/keystone/blob/da1359dfc1bff7e27505eff876efe3a0865bae2d/packages/keystone/lib/Keystone/index.js#L142-L144

Make the function param for `fastMemoize` into a promise
https://github.com/keystonejs/keystone/blob/da1359dfc1bff7e27505eff876efe3a0865bae2d/packages/keystone/lib/Keystone/index.js#L160-L165

This last one is a problem though because `fastMemoize` doesn't currently support promises like that. I'll check into another package. @jesstelford do you all have guidelines for dependencies as far as what is acceptable to use? Like number of contributors, recent releases, etc?

**Comment by wcalebgray** [comment]
Scratch that plan. I misread that `fastMemoize` was executing the function, when it's actually just saving it to use in other places within `context`. So...that means that all things that use the `getCustomAccessControlForUser` function from the `context` need to be async and on up that call stack. And I likely need to do this for not just this one, but for the other 3 access control resolvers as well. 😅

**Comment by wcalebgray** [comment]
@jesstelford opened a PR ⬆️ with those above changes. I probably missed some things (this is my first Keystone PR), so I'd welcome feedback on how to make it better/acceptable for Keystone!

**Comment by jesstelford** [comment]
It's merged 🎉
